cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,1014 @@
+			name:    "when the firewall cache is nil return existing firewall from API then update cache",
+			fwCache: newFakeFirewallCacheEmpty(),
+			firewallRequest: &godo.FirewallRequest{
+				Name:          testWorkerFWName,
+				InboundRules:  testInboundRules,
+				OutboundRules: testOutboundRules,
+				Tags:          testWorkerFWTags,
+			},
+			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return []godo.Firewall{*newFakeFirewall()}, newFakeOKResponse(), nil
+			},
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					listFunc:   test.expectedGodoFirewallListResp,
+					updateFunc: test.expectedGodoFirewallUpdateResp,
+					createFunc: test.expectedGodoFirewallCreateResp,
+					getFunc:    test.expectedGodoFirewallGetResp,
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, testWorkerFWTags, testWorkerFWName)
+
+			err := fc.fwManager.Set(ctx, test.firewallRequest)
+			if (err != nil && test.expectedError == nil) || (err == nil && test.expectedError != nil) {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", test.expectedError, err)
+			}
+		})
+	}
+}
+
+func TestFirewallController_NoDataRace(t *testing.T) {
+	// setup
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+	var wg sync.WaitGroup
+
+	gclient := newFakeGodoClient(
+		&fakeFirewallService{
+			listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			updateFunc: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(), newFakeOKResponse(), nil
+			},
+			createFunc: func(context.Context, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+				return nil, newFakeOKResponse(), nil
+			},
+			getFunc: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
+				return newFakeFirewall(), newFakeOKResponse(), nil
+			},
+		},
+	)
+	fwManagerOp := newFakeFirewallManagerOp(gclient, newFakeFirewallCache())
+	fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, testWorkerFWTags, testWorkerFWName)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for ctx.Err() == nil { // context has not been terminated
+			if err := fc.ensureReconciledFirewall(ctx); err != nil && err != context.Canceled {
+				t.Errorf("ensureReconciledFirewall failed: %s", err)
+			}
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		fc.Run(ctx.Done(), fwManagerOp, time.Duration(0))
+	}()
+
+	wg.Wait()
+	// We do not assert on anything because the goal of this test is to catch data races.
+}
+
+func TestFirewallController_actualRun(t *testing.T) {
```

Comment:
I'm not sure the test cases are perfectly aligned with the expectations we have for `actualRun`, at least naming-wise. I think we should look at the following scenarios and make sure we have tests for it:

1. An error is returned when we cannot get the firewall (i.e., `Get()` fails).
1. The firewall is not reconciled (i.e., `Set()` is not called) when the cache content matches the firewall we get.
1. The firewall is reconciled if the cache contents do not match.
1. An error is returned if the cache contents do not match but we fail to reconcile the firewall.

You could even use the `firewallManager` interface to stub out the `Get()` and `Set()` calls since we do not care about API details in this test. I'll leave it up to you to decide which interface to stub out.

Reply:
I ended up removing the FirewallManager interface and refactoring some things. But I stub out calls to the client and hit these edge cases in the actual run test @timoreimann

Reply:
```go
{
			name:    "error returns when we fail to get firewall from firewall api",
			fwCache: newFakeFirewallCacheEmpty(),
			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
				return nil, newFakeNotFoundResponse(), errors.New("not found")
			},
			expectedGodoFirewallListResp: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
				return nil, newFakeNotFoundResponse(), errors.New("not found")
			},
			expectedError: errors.New("failed to get worker firewall: failed to retrieve list of firewalls from DO API"),
		},
		{
			name:    "when firewall matches what we expect there is nothing to reconcile",
			fwCache: newFakeFirewallCache(),
			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
				return newFakeFirewall(), newFakeOKResponse(), nil
			},
		},
		{
			name:    "when firewall needs to be reconciled an update call is made",
			fwCache: newFakeFirewallCacheWithDiffInboundRules(),
			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
				return newFakeFirewall(), newFakeOKResponse(), nil
			},
			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
				return newFakeFirewallWithDiffInboundRules(), newFakeOKResponse(), nil
			},
		},
		{
			name:    "error is returned when firewall cache exists but needs to be reconciled and update call fails",
			fwCache: newFakeFirewallCacheWithDiffInboundRules(),
			expectedGodoFirewallGetResp: func(context.Context, string) (*godo.Firewall, *godo.Response, error) {
				return newFakeFirewall(), newFakeNotFoundResponse(), errors.New("not found")
			},
			expectedGodoFirewallUpdateResp: func(context.Context, string, *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
				return nil, newFakeNotOKResponse(), errors.New("unexpected error")
			},
			expectedError: errors.New("failed to reconcile worker firewall: failed to set reconciled firewall: could not update firewall: unexpected error"),
		},
```